File: python/modify_order.py
"""Place a resting order (GTC, 3% below mid), then modify its price to 4% below."""

import json
from client import exchange, sign_hash, get_mid

COIN = "BTC"

mid = get_mid(COIN)
if mid == 0:
    print(f"Could not fetch {COIN} mid price")
    raise SystemExit(1)

sz = round(11.0 / mid, 5)
rest_px = int(mid * 0.97)

print(f"{COIN} mid: ${mid:,.2f}")
print(f"Placing resting BUY {sz} @ {rest_px} (GTC, 3% below mid)\n")

res = exchange({
    "action": {
        "type": "order",
        "orders": [{
            "asset": COIN,
            "side": "buy",
            "price": str(rest_px),
            "size": str(sz),
            "tif": "gtc",
        }],
    },
})
sig = sign_hash(res["hash"])

result = exchange({
    "action": res["action"],
    "nonce": res["nonce"],
    "signature": sig,
})

exchange_resp = result["exchangeResponse"]
statuses = exchange_resp.get("response", {}).get("data", {}).get("statuses", [])

oid = None
for s in statuses:
    if isinstance(s, dict) and "resting" in s:
        oid = s["resting"].get("oid")
        break

if oid is None:
    print("Could not extract OID from resting order")
    print(json.dumps(exchange_resp, indent=2))
    raise SystemExit(1)

new_px = int(mid * 0.96)
print(f"Order resting (OID: {oid})")
print(f"Modifying price: {rest_px} -> {new_px}\n")

modify_action = {
    "type": "batchModify",
    "modifies": [{
        "oid": oid,
        "order": {
            "asset": COIN,
            "side": "buy",
            "price": str(new_px),
            "size": str(sz),
            "tif": "gtc",
        },
    }],
}

res = exchange({"action": modify_action})
sig = sign_hash(res["hash"])

result = exchange({
    "action": modify_action,
    "nonce": res["nonce"],
    "signature": sig,
})

print(json.dumps(result["exchangeResponse"], indent=2))
print("\nOrder modified.")

File: typescript/src/modifyOrder.ts
import { exchange, signHash, getMid } from "./client";

const COIN = "BTC";

async function main() {
  const mid = await getMid(COIN);
  if (mid === 0) {
    console.error(`Could not fetch ${COIN} mid price`);
    process.exit(1);
  }

  const sz = (11.0 / mid).toFixed(5);
  const restPx = Math.floor(mid * 0.97).toString();

  console.log(`${COIN} mid: $${mid.toLocaleString()}`);
  console.log(`Placing resting BUY ${sz} @ ${restPx} (GTC, 3% below mid)\n`);

  let res = await exchange({
    action: {
      type: "order",
      orders: [{ asset: COIN, side: "buy", price: restPx, size: sz, tif: "gtc" }],
    },
  });
  let sig = await signHash(res.hash);

  const result = await exchange({
    action: res.action,
    nonce: res.nonce,
    signature: sig,
  });

  const exchangeResp = result.exchangeResponse;
  const statuses = exchangeResp?.response?.data?.statuses || [];

  let oid: number | null = null;
  for (const s of statuses) {
    if (s.resting) {
      oid = s.resting.oid;
      break;
    }
  }

  if (oid === null) {
    console.error("Could not extract OID from resting order");
    console.log(JSON.stringify(exchangeResp, null, 2));
    process.exit(1);
  }

  const newPx = Math.floor(mid * 0.96).toString();
  console.log(`Order resting (OID: ${oid})`);
  console.log(`Modifying price: ${restPx} -> ${newPx}\n`);

  const modifyAction = {
    type: "batchModify",
    modifies: [{
      oid,
      order: { asset: COIN, side: "buy", price: newPx, size: sz, tif: "gtc" },
    }],
  };

  res = await exchange({ action: modifyAction });
  sig = await signHash(res.hash);

  const modifyResult = await exchange({
    action: modifyAction,
    nonce: res.nonce,
    signature: sig,
  });

  console.log(JSON.stringify(modifyResult.exchangeResponse, null, 2));
  console.log("\nOrder modified.");
}

main();
